storage/remote/queue_manager.go

```diff
@@ -946,10 +994,21 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue chan sample) {
 			// Number of pending samples is limited by the fact that sendSamples (via sendSamplesWithBackoff)
 			// retries endlessly, so once we reach max samples, if we can never send to the endpoint we'll
 			// stop reading from the queue. This makes it safe to reference pendingSamples by index.
-			pendingSamples[nPending].Labels = labelsToLabelsProto(sample.labels, pendingSamples[nPending].Labels)
-			pendingSamples[nPending].Samples[0].Timestamp = sample.t
-			pendingSamples[nPending].Samples[0].Value = sample.v
-			nPending++
+			switch d := sample.(type) {
+			case rwSample:
+				pendingSamples[nPending].Labels = labelsToLabelsProto(d.labels, pendingSamples[nPending].Labels)
+				pendingSamples[nPending].Samples[0].Timestamp = d.t
+				pendingSamples[nPending].Samples[0].Value = d.v
+				nPending++
+			case rwExemplar:
+				pendingSamples[nPending].Labels = labelsToLabelsProto(d.seriesLabels, pendingSamples[nPending].Labels)
+				pendingSamples[nPending].Exemplars = append(pendingSamples[nPending].Exemplars[:0], prompb.Exemplar{
+					Labels:    labelsToLabelsProto(d.labels, nil),
+					Value:     d.v,
+					Timestamp: d.t,
+				})
+				nPending++
```

Comment:
This would mean that we would now easily be repeating series, one with samples, and one with exemplars correct? If so, we would not be respecting the sample limit anymore, i.e. we would send significantly fewer samples per batch if exemplars are present.

Not sure if I have a problem with the duplication, but a quick discussion on pros/cons/concerns would be nice :)

Reply:
Yes. At the moment this PR considers exemplars to be samples for the purpose of tracking the # of pending samples. The storage interface is set up so that exemplars and samples are stored separately, so records are separate in the WAL as well. We can't reliably match up exemplars with a sample once this has happened, since exemplars most likely won't have the same timestamp as the sample they were scraped with.

We don't really do anything efficient right now in terms of grouping together of samples for the same series. My current running assumption is that even 1% of series having an exemplar on every scrape (that we end up actually storing and writing to the WAL) is going to be more than the average use case, so if that's the case this wouldn't impact overall throughput of samples much. The other option would to be start up s separate set of shards for exemplars data.

I also intend to put the remote write portion behind another feature flag.

Reply:
Yeah, I agree it is probably a very small number. If we need more efficiency later we can figure out grouping and such then. Maybe `pendingSamples` is just just misnamed a bit since it is actually a `[]prompb.TimeSeries`.

Reply:
are you suggesting a rename of the pending metric?